---
title: test_randoms
format:
     html:
         self-contained: true
         toc: true
         toc-location: left
         toc-expand: 2
---

# Basic Overview

## Theory

<h3>Rarefraction curves</h3>
<p>The x-axis of the plot shows the total number of sampled sequences, and the y-axis shows the total number of unique sequences.</p>
<p>The plot can be interpreted as follows:</p>
<ul>
<li>Samples with higher sequencing depth will have more unique sequences at each point on the x-axis.</li>
<li>Samples with lower sequencing depth will have fewer unique sequences at each point on the x-axis.</li>
</ul>
<p>This plot is useful for NGS because it can be used to assess the quality of the sequencing data and to determine whether the sequencing depth is sufficient for the desired application. For example, if you are trying to identify rare variants in a sample, you will need to have a high sequencing depth in order to be confident that the variants are real and not simply sequencing errors.</p>
<p>Here are some specific examples of how the plot could be used to interpret NGS data:</p>
<ol>
<li>If you are comparing two samples with different sequencing depths, you can use the plot to see which sample has more unique sequences at each point on the x-axis. This can help you to determine which sample is more likely to contain the variants of interest.</li>
<li>If you are trying to determine whether the sequencing depth of a sample is sufficient for a particular application, you can use the plot to compare the sample to other samples that have been used successfully for the same application.</li>
<li>If you are seeing a plateau in the number of unique sequences as the sequencing depth increases, this may indicate that you have reached a saturation point and that further sequencing will not yield much new information.</li>
</ol>

<h3>Alignment Quality Plot</h3>
<p>The barplot shows the total sequenced reads and aligned reads for each sample. The total sequenced reads is the number of reads that were generated by the NGS sequencer. The aligned reads is the number of reads that were successfully mapped to the reference genome.</p>
<p>The plot shows that the total sequenced reads is higher than the aligned reads for all samples. This is because some of the reads may be of low quality or may not map to the reference genome.</p>
<p>Furthermore, it shows that the difference between the total sequenced reads and aligned reads is larger for some samples than others. This may be due to a number of factors, such as the quality of the DNA sample, the sequencing platform used, and the alignment parameters used.</p>
<p>Use of the barplot for NGS:</p>
<p>The barplot can be used to assess the quality of the NGS data and to determine whether the alignment rate is sufficient for the desired application. </p>
<p>Specific examples:</p>
<ol>
<li>If you are comparing two samples, you can use the plot to see which sample has a higher alignment rate. This can help you to determine which sample is more likely to contain the variants of interest.</li>
<li>If you are trying to determine whether the alignment rate for a sample is sufficient for a particular application, you can compare the sample to other samples that have been used successfully for the same application.</li>
<li>If you are seeing a low alignment rate for a sample, this may indicate that there is a problem with the DNA sample, the sequencing platform used, or the mixcr parameters used.</li>
</ol>

<h3>Heatmap based on Morosita-Horn Index</h3>
<p>The heatmap shows a matrix which values are calculated based on the morosita horn index. This index captures the degree of identity between two samples which are composed of multiple values of different sizes. This is especially useful for comparing two samples which are composed of different numbers of clones and have a different number of aligned reads. In the context of this heatmap, identity means that two sequences have the same sequence length and the exact same amino acid sequence. </p>
<p>The heatmap has several possible applications during the quality control of the sequencing, which are:
1. identification of cross contamination between samples. This can be seen if a high degree of identity can be observed, although the samples were panned against different antigens.
2. Validation of the panning process. This can be seen if a high degree of identity can be observed between samples which were panned against the same antigen in different rounds.</p>
<p><em>Weaknesses</em>:
- the morosita horn becomes less accuracte if the number of elements (sequences) is very low in one of the samples</p>
<p><em>Alternatives</em>:</p>
<p>The user can analyse the identity between the samples based on:
- Jaccard Index
- Sorensen-Dice Index</p>

<h3>Logo Plot</h3>
<p>It shows the frequency of each amino acid at each position in the CDR3 region. The height of each letter in the logo represents the frequency of the amino acid at that position.</p>
<p>Sequence logo plots are useful for identifying conserved and variable regions in a protein sequence. Conserved regions are regions where the same amino acids are present in a high proportion of sequences. Variable regions are regions where different amino acids are present at a high frequency.</p>
<p>It is important to only choose sequences with the same length when creating a sequence logo plot because this allows for a more accurate comparison of the amino acid distribution at each position in the sequence. If sequences of different lengths are included in the plot, it can be difficult to interpret the results.</p>
<p>Here is an example of how the sequence logo plot you provided could be interpreted:</p>
<ol>
<li>The positions with the highest entropy are the most variable positions in the CDR3 region. </li>
<li>The positions with the lowest entropy are the most conserved positions in the CDR3 region. </li>
<li>The most common amino acids at each position are shown in the center of the logo.</li>
</ol>
<p>Sequence logo plots can be used for a variety of purposes, such as:</p>
<ol>
<li>Identifying conserved and variable regions in a protein sequence.</li>
<li>Identifying potential motifs in a protein sequence.</li>
<li>Comparing the amino acid distribution in two or more protein sequences.</li>
<li>Tracking the evolution of a protein sequence over time.</li>
</ol>

## Identity between samples

### Identity based on Morosita Horn Index

::: {layout-ncol=1}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\morosita_horn.png)

:::

### Identity based on Jaccard Index

::: {layout-ncol=1}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\jaccard.png)

:::

{{< pagebreak >}}



### Identity based on Sorensen Index

{{< pagebreak >}}



::: {layout-ncol=1}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sorensen.png)

:::

{{< pagebreak >}}



## Sequencing Quality of samples

::: {layout-ncol=2}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\alignment_quality.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\rarefraction_all.png)

General Sequencing Quality

:::

{{< pagebreak >}}



## Sample specific quality analysis

### Quality analysis for random1_R1_001

::: {layout="[[1, 1], [1]]"}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\length_distributions\random1_R1_001.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\rarefraction_curves\rarefraction_curves_random1_R1_001.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\logo_plots\random1_R1_001.png)

:::

{{< pagebreak >}}



### Quality analysis for random2_R1_001

::: {layout="[[1, 1], [1]]"}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\length_distributions\random2_R1_001.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\rarefraction_curves\rarefraction_curves_random2_R1_001.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\logo_plots\random2_R1_001.png)

:::

{{< pagebreak >}}



### Quality analysis for random3_1_001

::: {layout="[[1, 1], [1]]"}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\length_distributions\random3_1_001.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\rarefraction_curves\rarefraction_curves_random3_1_001.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\logo_plots\random3_1_001.png)

:::

{{< pagebreak >}}



### Quality analysis for random4_R1_001

::: {layout="[[1, 1], [1]]"}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\length_distributions\random4_R1_001.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\rarefraction_curves\rarefraction_curves_random4_R1_001.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\logo_plots\random4_R1_001.png)

:::

{{< pagebreak >}}



# Sequence Clusters

## Theory

<h3>Clustering based on levenshtein distance</h3>
<p>The Levenshtein distance is a metric that measures the similarity between two strings. It is calculated by counting the minimum number of edits (insertions, deletions, and substitutions) required to transform one string into the other.</p>
<p>For example, the Levenshtein distance between the strings "CAT" and "DOG" is 2, because we need to insert the letter "D" and substitute the letter "C" with the letter "G" to transform "CAT" into "DOG".</p>
<p>The Levenshtein distance can be used to analyze peptide sequences in a variety of ways. For example, it can be used to:</p>
<ol>
<li>Identify similar CDR3 regions. This can be useful for identifying antibodies with similar antigen-binding specificities.</li>
<li>Detect mutations in CDR3 regions. This can be useful for identifying mutations that may affect the affinity or specificity of an antibody.</li>
<li>Cluster CDR3 regions into groups. This can be useful for identifying groups of CDR3 regions with similar antigen-binding specificities.</li>
</ol>
<p>The connected components in the plot (shown by a a network of lines) are groups of CDR3 sequences that are similar to each other, based on a threshold for a levenshtein distance (default = 2). The size of each circle in the plot represents the number of clones for the corresponding CDR3 sequence relative to the clone counts of other sequences.</p>
<p>Overall, the plot can be used to get a general overview of the diversity and similarity of the CDR3 sequences in the library.</p>
<p>Furthermore, a table is given which has the sorted sequences based on the clone count. That means the top 10 sequences represent the sequences with the highest clone fraction in that sample. This table, coming from a report can be used to identify specificity for a certain antigen based on a density of these sequences in certain clusters.</p>
<p>Disadvantages of the levenshtein distance</p>
<ul>
<li>Only captures global similarity of sequences and does not focues on specific regions</li>
<li>Does not take any chemical properties or structural properties into account</li>
</ul>

<h3>Sequence Cluster Dendrogram</h3>
<p>The dendrogram you provided is a visualization of the similarity between a set of CDR3 heavy chain sequences. The sequences are clustered together based on their Levenshtein distance, which is described in the previous section.</p>
<p>The dendrogram can be used to understand the diversity of the CDR3 sequences in a sample. For example, if the dendrogram shows that most of the sequences are clustered together in a few large clusters, this suggests that there is a relatively low level of diversity in the sample. On the other hand, if the dendrogram shows that the sequences are spread out in many small clusters, this suggests that there is a high level of diversity in the sample.</p>
<p>The dendrogram can also be used to identify groups of similar sequences. For example, if we want to identify a group of sequences that are likely to have similar antigen-binding specificities, we can look for a cluster of sequences that are close to each other on the dendrogram.</p>

<h3>Clustering based on t-SNE</h3>
<p>Addressing the limitations of using the levenshtein distance for clustering hcdr3 sequences, a plot which shows the sequences and their whole sequence relation was created to enable a more extensive analysis of the sequences and their binding. To be able to have the sequences as vectors the SGT embedding was applied.</p>
<p>This embedding does not capture any chemical properties of the amino acid strand but instead tries to recognize the characteristic relative position of letters within a sequence which enables an identification of patterns between sequences of different lengths. This addresses the variability of the cdr3 sequences and enables to capture the similarity of sequences. Therefore, the sgt embedder was initialized using the package sgt where the parameter length sensitive was set to True and the parameter kappa was assigned with 1 . After the initialization, the embedding creates  an output of 400 dimensions per sequence which were reduced to 80 dimensions by using principal component analysis (PCA).</p>
<p>Subsequently t-SNE was implemented to show a representative arrangement of the 80 dimensions in the two dimensional space because a further dimension reduction to only two dimensions would insufficiently describe the data globally by using PCA. That is why t-SNE was introduced to be able to show a representative arrangement of the points without losing too much of its information. </p>
<p>The plot can be used in variety of ways, such as: 
1. Identify similar CDR3 regions
2. learn about the differences between samples
3. identify specific sequence differences for samples which were panned against different antigens.</p>

## Levenshtein distance based

### Sequence Similarity based on Levenshtein Distance of random1_R1_001

::: {layout-ncol=2}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_cluster\random1_R1_001ls_cluster.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_cluster\random1_R1_001ls_dendro.png)



:::

|    |   Unnamed: 0 |   Cluster No. | Sequences            |
|---:|-------------:|--------------:|:---------------------|
|  0 |            0 |             0 | CNAVHSRWQAMTRW       |
|  1 |            1 |             1 | CIQSGTDRR            |
|  2 |            2 |            30 | CNVDLTVVDGRHLPRGDYW  |
|  3 |            3 |             1 | CIQSGTSRR            |
|  4 |            4 |             9 | CHADLRVRDGVRGDYW     |
|  5 |            5 |             0 | CNAVHSRWQAMTHW       |
|  6 |            6 |             4 | CAADLFGTRQADLLIYNFR  |
|  7 |            7 |             5 | CNAVGADRISGVIYW      |
|  8 |            8 |             6 | CATRFTTPWDARAPAYYNYW |
|  9 |            9 |             7 | CAAQVYTSGIYYYSGSYDYW |

{{< pagebreak >}}



### Sequence Similarity based on Levenshtein Distance of random2_R1_001

::: {layout-ncol=2}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_cluster\random2_R1_001ls_cluster.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_cluster\random2_R1_001ls_dendro.png)



:::

|    |   Unnamed: 0 |   Cluster No. | Sequences                  |
|---:|-------------:|--------------:|:---------------------------|
|  0 |            0 |             4 | CAADSPSLPDRTAAHSYDYAYW     |
|  1 |            1 |             1 | CAAERVPSAIYTHYEFCVGADEYDYW |
|  2 |            2 |             2 | CAADLFGTRQADLLIYNFR        |
|  3 |            3 |             3 | CAAGSDYNLDGSW              |
|  4 |            4 |             4 | CAAESPDLVDRTATHSYDYSYW     |
|  5 |            5 |             5 | CHADIRVRAGVRGDYW           |
|  6 |            6 |             6 | CGADKFPYSAETMCSIPGGPDLDAW  |
|  7 |            7 |             4 | CAAESPFLPDRTARHSYDYTYW     |
|  8 |            8 |             7 | CNAVHSRWQAMTRW             |
|  9 |            9 |             8 | CAADKFPYAAETMCVIPGGPDADTW  |

{{< pagebreak >}}



### Sequence Similarity based on Levenshtein Distance of random3_1_001

::: {layout-ncol=2}



:::

{{< pagebreak >}}



### Sequence Similarity based on Levenshtein Distance of random4_R1_001

::: {layout-ncol=2}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_cluster\random4_R1_001ls_cluster.png)

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_cluster\random4_R1_001ls_dendro.png)



:::

|    |   Unnamed: 0 |   Cluster No. | Sequences                  |
|---:|-------------:|--------------:|:---------------------------|
|  0 |            0 |             7 | CAADSPSLPDRTAAHSYDYAYW     |
|  1 |            1 |             1 | CNAVHSRWQAMTRW             |
|  2 |            2 |             2 | CIQSGTSRR                  |
|  3 |            3 |             2 | CIQSGTDRR                  |
|  4 |            4 |             1 | CNAVHSRWQAMTHW             |
|  5 |            5 |             3 | CNTEEESTGTYYEW             |
|  6 |            6 |             7 | CAAESPDLVDRTATHSYDYSYW     |
|  7 |            7 |             5 | CAAERVPSAIYTHYEFCVGADEYDYW |
|  8 |            8 |             6 | CAAGSDYNLDGSW              |
|  9 |            9 |             2 | CIQSGTDRK                  |

{{< pagebreak >}}



## Sequence Similarity between samples

### Global Sequence similarity for random1_R1_001

::: {layout-ncol=1}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_embedding\random1_R1_001embedding_tsne.png)

:::

### Global Sequence similarity for random2_R1_001

::: {layout-ncol=1}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_embedding\random2_R1_001embedding_tsne.png)

:::

### Global Sequence similarity for random3_1_001

::: {layout-ncol=1}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_embedding\random3_1_001embedding_tsne.png)

:::

### Global Sequence similarity for random4_R1_001

::: {layout-ncol=1}

![](C:\Users\nilsh\my_projects\ExpoSeq\my_experiments\test_randoms\plots\aaSeqCDR3\sequence_embedding\random4_R1_001embedding_tsne.png)

:::